truffle/src/main/ruby/core/proc.rb

```diff
@@ -33,18 +33,41 @@ class Proc
 
   def curry(curried_arity = nil)
     if lambda? && curried_arity
+
       if arity >= 0 && curried_arity != arity
         raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
           curried_arity,
           arity
         ]
       end
 
-      if arity < 0 && curried_arity < (-arity - 1)
-        raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
-          curried_arity,
-          -arity - 1
-        ]
+      if arity < 0
+        total_args = 0
+        req_args = 0
+        has_rest = false
+
+        for p in parameters do
+          type = p.first
+          req_args += (type == :req ? 1 : 0)
+          total_args += (type != :block ? 1 : 0)
+          has_rest |= type == :rest 
+        end
+      
+        if (curried_arity < req_args) ||
+           (!has_rest && (curried_arity > total_args))
+          if has_rest
+            raise ArgumentError, "Wrong number of arguments (given %i, expected %i+)" % [
+              curried_arity,
+              req_args
+            ]
```

Comment:
Again is this from Rubinius? I'd normally write `"... (given #{curried_arity}, expected #{req_args}+)"`.